pandas/core/reshape/reshape.py

```diff
@@ -481,15 +487,27 @@ def _unstack_frame(obj, level, fill_value=None):
             new_placement = new_columns.get_indexer(new_items)
             new_values, mask = bunstacker.get_new_values()
 
-            mblk = make_block(mask.T, placement=new_placement)
-            mask_blocks.append(mblk)
+            mask_blocks[new_placement] = mask.any(0)
 
-            newb = make_block(new_values.T, placement=new_placement)
-            new_blocks.append(newb)
-
-        result = DataFrame(BlockManager(new_blocks, new_axes))
-        mask_frame = DataFrame(BlockManager(mask_blocks, new_axes))
-        return result.loc[:, mask_frame.sum(0) > 0]
+            # BlockManager can't handle SparseBlocks with multiple items,
```

Comment:
I would like to move this logic to a function in internals. e.g. you define ``BlockManager.unstack``, then you have an unstack on Block. This way the override logic is pretty straightforward as its done on a block basis, and this is setup so that you can return multiple blocks (in case of sparse) from these routines.

you can pass the ``_Unstacker`` as a partial function, and so the entire loop looks something like

``obj._data.unstack(_Unstacker)``

then call the ``_Unstacker`` inside the ``block.unstack()`` method.